contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/Batch.java

```diff
@@ -0,0 +1,96 @@
+package com.linkedin.parseq.batching;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+import com.linkedin.parseq.batching.BatchImpl.BatchBuilder;
+import com.linkedin.parseq.batching.BatchImpl.BatchEntry;
+import com.linkedin.parseq.promise.Promise;
+import com.linkedin.parseq.promise.PromiseResolvedException;
+import com.linkedin.parseq.promise.SettablePromise;
+
+/**
+ * Batch represents a collection of keys related to each other in such a way that
+ * is more efficient to compute values for those keys in bulk than computing value
+ * for each key individually.
+ * This class contains methods helpful in implementing bulk operation that completes
+ * Promises associated with keys.
+ *
+ * @author Jaroslaw Odzga (jodzga@linkedin.com)
+ *
+ * @param <K> Type of a Key
+ * @param <T> Type of a Value
+ */
+public interface Batch<K, T> {
+
+  /**
+   * Performs the given action for each element of the batch
+   * until all elements have been processed or the action throws an
+   * exception. Order in which elements are processed is unspecified.
+   * Exceptions thrown by the action are relayed to the caller.
+   * @param consumer action to be performed in each element of the batch
+   */
+  void foreach(BiConsumer<K, SettablePromise<T>> consumer);
+
+  /**
+   * Returns set of keys belonging to this batch.
+   * @return set of keys belonging to this batch
+   */
+  Set<K> keys();
+
+  /**
+   * Returns size of this batch.
+   * @return size of this batch.
+   */
+  int size();
+
+  /**
+   * Completes a {@link Promise} associated with given key with
+   * a value.
+   * Throws PromiseResolvedException if Promise associated with given key has already been resolved.
+   * @param key key that identifies a Promise to be completed
+   * @param value value to complete Promise with
+   * @throws PromiseResolvedException if Promise associated with given key has already been resolved
+   */
+  void done(K key, T value) throws PromiseResolvedException;
+
+
+  /**
+   * Fails a {@link Promise} associated with given key with
+   * a an error.
+   * Throws PromiseResolvedException if Promise associated with given key has already been resolved.
+   * @param key key that identifies a Promise to be completed
+   * @param error error to fail Promise with
+   * @throws PromiseResolvedException if Promise associated with given key has already been resolved
+   */
+  void fail(K key, Throwable error) throws PromiseResolvedException;
+
+  /**
+   * Fails all promises belonging to this batch with given error.
+   * This method throws PromiseResolvedException if any of the promises
+   * have already been resolved.
+   * This method guarantees that each Promise in this batch is either completed or failed.
```

Comment:
This could be made more clear, all remaining promises will be marked as failed even if one of them throws a PromiseResolvedException?